src/wallet/rpcdump.cpp

```diff
@@ -549,6 +549,57 @@ UniValue dumpwallet_impl(const UniValue& params, bool fHelp, bool fDumpZKeys)
     return exportfilepath.string();
 }
 
+class AddSpendingKeyToWallet : public boost::static_visitor<bool>
+{
+private:
+    CWallet *m_wallet;
+public: 
+    AddSpendingKeyToWallet(CWallet *wallet) : m_wallet(wallet) {}
+
+    bool operator()(const libzcash::SproutSpendingKey &sk) const {
+        auto addr = sk.address();
+        // Don't throw error in case a key is already there
+        if (m_wallet->HaveSpendingKey(addr)) {
+            return true;
+        } else {
+            m_wallet->MarkDirty();
+
+            if (!m_wallet-> AddZKey(sk)) {
+                throw JSONRPCError(RPC_WALLET_ERROR, "Error adding spending key to wallet");
+            }
+
+            m_wallet->mapZKeyMetadata[addr].nCreateTime = 1;
+            
+            return false;
+        }
+    }
+    
+    bool operator()(const libzcash::SaplingSpendingKey &sk) const {
+        auto fvk = sk.full_viewing_key();
+        auto addr = sk.default_address();
+        {
+            // Don't throw error in case a key is already there
+            if (m_wallet->HaveSaplingSpendingKey(fvk)) {
+                return true;
+            } else {
+                m_wallet->MarkDirty();
+    
+                if (!m_wallet-> AddSaplingZKey(sk)) {
+                    throw JSONRPCError(RPC_WALLET_ERROR, "Error adding spending key to wallet");
+                }
+    
+                m_wallet->mapSaplingZKeyMetadata[addr].nCreateTime = 1;
+                
+                return false;
+            }    
+        }
+    }
+    
+    bool operator()(const libzcash::InvalidEncoding& no) const { return true; }
```

Comment:
This is probably fine; or you could throw an error here (this case should never be triggered, as we call `IsValidSpendingKey()` first which ensures that `spendingkey` is not an `InvalidEncoding`, so getting here would be an implementation error).